core/src/main/scala/quasar/lib/jdbc/JdbcDiscovery.scala

```diff
@@ -229,6 +90,162 @@ object JdbcDiscovery {
+      * 22. SOURCE_DATA_TYPE short => source type of a distinct type or user-generated Ref type, SQL type from java.sql.Types (null if DATA_TYPE isn't DISTINCT or user-generated REF)
+      * 23. IS_AUTOINCREMENT String => Indicates whether this column is auto incremented
+      *     YES --- if the column is auto incremented
+      *     NO --- if the column is not auto incremented
+      *     empty string --- if it cannot be determined whether the column is auto incremented
+      * 24. IS_GENERATEDCOLUMN String => Indicates whether this is a generated column
+      *    YES --- if this a generated column
+      *    NO --- if this not a generated column
+      *    empty string --- if it cannot be determined whether this is a generated column
+      */
+    private implicit val columnMetaRead: Read[ColumnMeta] =
+      new Read[ColumnMeta](Nil, (rs, _) =>
+        ColumnMeta(
+          ColumnName(rs.getString(4)),
+          JdbcType.fromInt(rs.getInt(5)),
+          rs.getString(6),
+          ColumnNullable.fromInt(rs.getInt(11)) getOrElse ColumnNullable.NullableUnknown))
+
+    private def selectedColumns(selector: TableSelector): Stream[ConnectionIO, ColumnMeta] = {
+      val (schemaPattern, tablePattern) = selector
+
+      Stream.force(for {
+        catalog <- HC.getCatalog
+        rs <- HC.getMetaData(FDMD.getColumns(catalog, schemaPattern, tablePattern, "%"))
+        ts = HRS.stream[ColumnMeta](JdbcDiscovery.MetaChunkSize)
+      } yield ts.translate(λ[ResultSetIO ~> ConnectionIO](FC.embed(rs, _))))
+    }
+
+    private def selectedTables(selector: TableSelector): Stream[ConnectionIO, TableMeta] = {
+      val (schemaPattern, tablePattern) = selector
+
+      Stream.force(for {
+        catalog <- HC.getCatalog
+        types <- discoverableTableTypes.sequence
+        typeMask = types.map(_.map(_.name).toSortedSet.toArray)
+        rs <- HC.getMetaData(FDMD.getTables(catalog, schemaPattern, tablePattern, typeMask.orNull))
+        ts = HRS.stream[TableMeta](JdbcDiscovery.MetaChunkSize)
+      } yield ts.translate(λ[ResultSetIO ~> ConnectionIO](FC.embed(rs, _))))
+    }
+
+    private def tableSelector(ref: SchemaName Ior TableName): ConnectionIO[TableSelector] =
+      HC.getMetaData(FDMD.getSearchStringEscape) map { escape =>
+        ref match {
+          // All tables in the schema
+          case Ior.Left(schema) =>
+            (escapeForSearch(escape, schema.asString), "%")
+
+          // The table having no schema
+          case Ior.Right(table) =>
+            ("", escapeForSearch(escape, table.asString))
+
+          // The table in the specified schema
+          case Ior.Both(schema, table) =>
+            (escapeForSearch(escape, schema.asString), escapeForSearch(escape, table.asString))
+        }
+      }
+
+    private def escapeForSearch(escapeString: String, patternLiteral: String): String = {
```

Comment:
We should have tests to enforce that this kind of logic works as expected and keeps working as expected in all concrete dbms-es.

Not sure if we need to do this now (and for v2), but good to keep in mind for v3 at least.

Reply:
I'd vote to put it as is for now, and refactor it when we need for v3. I.e. extract escape code somewhere and make `selectedTables` not working without it, like 
```scala
trait Escaper[+F[_]] {
  def apply(i: String): F[String]
}
object Escaper {
  def jdbc: Escaper[ConnectionIO] = ???
  def apply(escape: String): Escaper[Nothing] = ???
}
// in JdbcDiscovery
def apply(discoverableTypes: ..., escaper: Escaper[ConnectionIO]): ConnectionIO
```